### PR TITLE
Deprecate `serialiseTxLedgerCddl`

### PR DESCRIPTION
**Pull request by palas**
# Changelog

```yaml
- description: |
    Added deprecation warning to function `serialiseTxLedgerCddl`
  type:
   - breaking
```
# Context

In the past, we had an intermediate tx body format (sometimes called the intermediate cli format). This format did not adhere to the CDDL format in the ledger. So `serialiseTxLedgerCddl`  was introduced to remedy this, with the aim of dropping the "intermediate cli format".

We have completely dropped the "intermediate cli format" so both functions will serialize txs that adhere to the ledger's CDDL format.

Therefore `serialiseTxLedgerCddl` is no longer necessary.

This PR deprecates it and reimplements `serialiseTxLedgerCddl` to use `serialiseToTextEnvelope`, which causes slight changes to the formats, see `How to trust this PR` below for more detail.

## `cardano-api`

- Main PR: https://github.com/IntersectMBO/cardano-api/pull/534
- Branch: [cardano-api/deprecate-serialiseTxLedgerCddl](https://github.com/IntersectMBO/cardano-api/tree/deprecate-serialiseTxLedgerCddl)
- Backported branch: [cardano-api/deprecate-serialiseTxLedgerCddl-testing](https://github.com/IntersectMBO/cardano-api/tree/deprecate-serialiseTxLedgerCddl-testing)

## `cardano-cli`

- PR: https://github.com/IntersectMBO/cardano-cli/pull/772
- Branch: [cardano-cli/deprecate-serialiseTxLedgerCddl-testing](https://github.com/IntersectMBO/cardano-cli/tree/deprecate-serialiseTxLedgerCddl-testing)

## `cardano-node`

- PR: https://github.com/IntersectMBO/cardano-node/pull/5867
- Branch: [cardano-node/deprecate-serialiseTxLedgerCddl-testing](https://github.com/IntersectMBO/cardano-node/tree/deprecate-serialiseTxLedgerCddl-testing)

# How to trust this PR

Make sure that the message makes sense, that the functions deprecated are the one we want to deprecate, and that there are no other functions associated that we want to deprecate. Also that we want to override the deprecation in the tests (which makes sense to me, but maybe we want to remove or adapt the tests instead).

Also make sure the amount of backwards compatibility adaptors is adequate and in the right places.

Migrating to using `serialiseToTextEnvelope` changes the serialization format in three ways:
- `description`: is provided by current functions but is a parameter to `serialiseToTextEnvelope`:
  * This is not a problem for `deserialiseFromTextEnvelope` because it ignores the field, so we get compatibility for free.
  * We pass the old descriptions to `serialiseToTextEnvelope` to ensure old functions keep producing the same `description` just to avoid breaking golden tests at this point.
 - `type`: is deduced from the `HasTextEnvelope` instance of the serialised type, but `serialiseTxLedgerCddl` has hardcoded types which do not match `HasTextEnvelop` instances:
   * This PR modified `serialiseToTextEnvelope` to loosely compare types (see `legacyComparison` function) allowing for support for the `serialiseTxLedgerCddl` versions.
   * We keep the old types in `serialiseTxLedgerCddl` by wrapping `serialiseToTextEnvelope` and patch the output to produce the old types too. This is, again, just to avoid breaking golden tests at this point.
- `rawCBOR`: is serialised pretty much identically by `serialiseToTextEnvelope` with one exception:
  * `serialiseWitnessLedgerCddl` encodes `KeyWitness` slightly differently than `serialiseToTextEnvelope`. In fact, `deserialiseWitnessLedgerCddl` relies on `description` to decide on the constructor to deserialise to of `KeyWitness`. This is a breaking change in the format and is not parsable by `deserialiseFromTextEnvelope` because it ignores the `description`. Changing `deserialiseFromTextEnvelope` to support this would be very patchy. So, this PR:
    - Modifies `serialiseWitnessLedgerCddl` to directly use `serialiseToTextEnvelope` (while keeping descriptions).
    - Modifies `deserialiseWitnessLedgerCddl` to use `deserialiseFromTextEnvelope` and to use the old decoding when that fails (this is done by the sub-function `legacyDecoding`).

These decisions provide a fair level of backwards compatibility but, at the same time, CBOR produced after the PR will be compatible with `deserialiseFromTextEnvelope` so functions in `SerialiseLedgerCddl` will become less necessary over time.

Also, to check these claims, you can see the results of running the tests in `cardano-testnet` and `cardano-cli` that are in the PRs linked in the `Context` section above.

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated. See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [x] Self-reviewed the diff

### CODE REVIEW

File: cardano-api/test/cardano-api-test/Test/Cardano/Api/Typed/CBOR.hs [hunk 1]
```diff
@@ -1,3 +1,4 @@
+{-# OPTIONS_GHC -Wno-deprecations #-}
```
Comment (smelc) [inline cardano-api/test/cardano-api-test/Test/Cardano/Api/Typed/CBOR.hs:1]:
```suggestion
{-# OPTIONS_GHC -Wno-deprecations #-} -- TODO remove when serialiseTxLedgerCddl is removed
```

**Review by smelc — APPROVED** [review]

File: cardano-api/internal/Cardano/Api/SerialiseLedgerCddl.hs [hunk 2]
```diff
@@ -134,6 +134,7 @@ instance Error TextEnvelopeCddlError where
     TextEnvelopeCddlErrByronKeyWitnessUnsupported ->
       "TextEnvelopeCddl error: Byron key witnesses are currently unsupported."
 
+{-# DEPRECATED serialiseTxLedgerCddl "Use 'serialiseToTextEnvelope' from 'Cardano.Api.SerialiseTextEnvelope' instead." #-}
 serialiseTxLedgerCddl :: ShelleyBasedEra era -> Tx era -> TextEnvelopeCddl
```
Comment (Jimbo4350) [inline cardano-api/internal/Cardano/Api/SerialiseLedgerCddl.hs:138]: We can go further here. We can have `serialiseTxLedgerCddl` call `serialiseToTextEnvelope` and remove some of the additional unnecessary types e.g `TextEnvelopeCddl`

**Review by Jimbo4350 — COMMENTED** [review]
Looking good but we should take this a little further.

File: cardano-api/test/cardano-api-test/Test/Cardano/Api/Typed/CBOR.hs (hunk 1 above)
Comment (palas) [reply cardano-api/test/cardano-api-test/Test/Cardano/Api/Typed/CBOR.hs]: Fixed in: https://github.com/IntersectMBO/cardano-api/pull/534/commits/a1766631208aa270b21c47e3e89fd2f128cb27e9

File: cardano-api/internal/Cardano/Api/SerialiseLedgerCddl.hs (hunk 2 above)
Comment (palas) [reply cardano-api/internal/Cardano/Api/SerialiseLedgerCddl.hs]: Like in [this commit](https://github.com/IntersectMBO/cardano-api/pull/534/commits/b773202f655d9918e2891af7195308639cd9f1fe)?

It looks like this is likely a breaking change. Is that ok?

**Review by carbolymer — COMMENTED** [review]
LGTM. Can you make sure that tests in cardano-testnet and cardano-cli pass prior to merge?

File: cardano-api/internal/Cardano/Api/SerialiseTextEnvelope.hs [hunk 3]
```diff
@@ -159,9 +160,29 @@ instance Error TextEnvelopeError where
...
 expectTextEnvelopeOfType :: TextEnvelopeType -> TextEnvelope -> Either TextEnvelopeError ()
 expectTextEnvelopeOfType expectedType TextEnvelope { teType = actualType } =
-    unless (expectedType == actualType) $
+    unless (expectedType `legacyComparison` actualType) $
       Left (TextEnvelopeTypeError [expectedType] actualType)
 
+-- | This is a backwards-compatibility patch to ensure that old envelopes
+-- generated by 'serialiseTxLedgerCddl' can be deserialised after switching
+-- to the 'serialiseToTextEnvelope'.
+legacyComparison :: TextEnvelopeType -> TextEnvelopeType -> Bool
```
Comment (Jimbo4350) [inline cardano-api/internal/Cardano/Api/SerialiseTextEnvelope.hs:169]: Nice :+1: . What we want to do is complain in the cli when a "legacy" text envelope type is used. The complaint would instruct them to convert the `type` field to the "correct" type. However before we do this, we should decide on a type format. I.e should we use the `Witnesses` vs `Unwitnessed` distinction or use the previous `Tx $era` format? You could start a discussion in discord on this.

**Review by Jimbo4350 — COMMENTED** [review]
Nice work :+1:

**Pull request merged**